esquery.js

```diff
@@ -47,196 +46,267 @@ function getPath(obj, key) {
+                                this.break();
+                                return;
+                            }
                         }
                     },
                     leave () { a.shift(); },
                     keys: options && options.visitorKeys,
                     fallback: options && options.fallback || 'iteration'
                 });
-            }
-            return collector.length !== 0;
 
+                return result;
+            };
         }
         case 'child':
-            if (matches(node, selector.right, ancestry, options)) {
-                return matches(ancestry[0], selector.left, ancestry.slice(1), options);
-            }
-            return false;
+            return (node, ancestry, options) => {
+                if (matches(node, selector.right, ancestry, options)) {
+                    return matches(ancestry[0], selector.left, ancestry.slice(1), options);
+                }
+                return false;
+            };
 
         case 'descendant':
-            if (matches(node, selector.right, ancestry, options)) {
-                for (let i = 0, l = ancestry.length; i < l; ++i) {
-                    if (matches(ancestry[i], selector.left, ancestry.slice(i + 1), options)) {
-                        return true;
+            return (node, ancestry, options) => {
+                if (matches(node, selector.right, ancestry, options)) {
+                    for (let i = 0, l = ancestry.length; i < l; ++i) {
+                        if (matches(ancestry[i], selector.left, ancestry.slice(i + 1), options)) {
+                            return true;
+                        }
                     }
                 }
-            }
-            return false;
+                return false;
+            };
 
         case 'attribute': {
-            const p = getPath(node, selector.name);
+            const path = selector.name.split('.');
             switch (selector.operator) {
                 case void 0:
-                    return p != null;
+                    return (node) => getPath(node, path) != null;
                 case '=':
                     switch (selector.value.type) {
-                        case 'regexp': return typeof p === 'string' && selector.value.value.test(p);
-                        case 'literal': return `${selector.value.value}` === `${p}`;
-                        case 'type': return selector.value.value === typeof p;
+                        case 'regexp': return (node) => {
+                            const p = getPath(node, path);
+                            return typeof p === 'string' && selector.value.value.test(p);
+                        };
+                        case 'literal': {
+                            const literal = `${selector.value.value}`;
+                            return (node) => {
+                                return literal === `${getPath(node, path)}`;
+                            };
```

Comment:
Style nit: you can use a concise arrow body here and below.

```suggestion
                            return (node) => literal === `${getPath(node, path)}`
```